core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4651,7 +4651,12 @@ public RexNode register(
               currentProjection);
         }
 
-        setRoot(newLeftInput, false);
+        // keep leaves property.
+        if (leaves.remove(root()) != null) {
+          setRoot(newLeftInput, true);
+        } else {
+          setRoot(newLeftInput, false);
+        }
```

Comment:
What about compacting into this equivalent version?

```suggestion
        setRoot(newLeftInput, leaves.remove(root()) != null);
```

Reply:
Good, Thanks asolimando!